The lastValue should be stored GC counted

maraisr opened the pull request:
#### 16249cfd52262119a7c9b0c28a74daf6f5910054
<pre>
The lastValue should be stored GC counted
<a href="https://bugs.webkit.org/show_bug.cgi?id=284334">https://bugs.webkit.org/show_bug.cgi?id=284334</a>

Reviewed by Darin Adler.

The lastValue is held on, possibly between GC phases
as such we should ensure we visit the value.

* Source/WebCore/dom/InternalObserverLast.cpp:

Canonical link: <a href="https://commits.webkit.org/287715@main">https://commits.webkit.org/287715@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c61e2385573c4a5a2df31da447e204d45898cfb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80189 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59191 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33602 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84706 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31165 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82300 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7490 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62684 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20499 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83258 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52753 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73028 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42990 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50083 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27183 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29626 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71212 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86138 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7409 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5243 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70958 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7584 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68868 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70198 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14192 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13142 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12482 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7373 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12890 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7212 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10732 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9017 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -53,45 +54,44 @@ class InternalObserverLast final : public InternalObserver {
+        if (UNLIKELY(!m_lastValue))
+            return protectedPromise()->reject(Exception { ExceptionCode::RangeError, "No values in Observable"_s });
 
-        if (UNLIKELY(!lastValue))
-            return Ref { m_promise }->reject(Exception { ExceptionCode::RangeError, "No values in Observable"_s });
-
-        Ref { m_promise }->resolve<IDLAny>(*lastValue);
+        protectedPromise()->resolve<IDLAny>(m_lastValue.getValue());
```
This changes behavior. The old code cleared out `m_lastValue` before calling `resolve`. is that no longer valuable?

maraisr replied:
I don't think we need to worry about that anymore. What I wanted to do with the clear out, was to no longer hold a reference to the value. Which was all wrong assumptions with how JSValue works. The JSValueInWrappedObject is much better, and by not visiting it anymore (as [mentioned here](https://github.com/WebKit/WebKit/pull/37670/files#r1876994747)), is doing the clearing out for us.

maraisr commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -53,45 +54,44 @@ class InternalObserverLast final : public InternalObserver {
-            return Ref { m_promise }->reject(Exception { ExceptionCode::RangeError, "No values in Observable"_s });
-
-        Ref { m_promise }->resolve<IDLAny>(*lastValue);
+        protectedPromise()->resolve<IDLAny>(m_lastValue.getValue());
```
I don't think it matters, given once we `complete`, the `visitAdditionalChildren` will stop being ran, as the m_active flag gets set to false upstream.

See: 
https://github.com/WebKit/WebKit/blob/3ff22d64c7b7ae1028986cac486f581e05485bff/Source/WebCore/dom/InternalObserver.h#L46-L49

https://github.com/WebKit/WebKit/blob/3ff22d64c7b7ae1028986cac486f581e05485bff/Source/WebCore/dom/InternalObserver.h#L65

But is it worth calling `.clear` on the value?

```suggestion
        protectedPromise()->resolve<IDLAny>(m_lastValue.getValue());
        
        m_lastValue.clear();
```

darinadler commented on Source/WebCore/dom/InternalObserverLast.cpp:
```diff
@@ -53,45 +54,44 @@ class InternalObserverLast final : public InternalObserver {
-    void visitAdditionalChildren(JSC::AbstractSlotVisitor&) const final
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor& visitor) const final
     {
+        m_lastValue.visit(visitor);
```
Is this safe when `m_lastValue` is nullptr? Does it crash in this case?

maraisr replied:
It is now a `JSValueInWrappedObject`, which is initialized as an empty class. Calling `.visit` does check internally for the presence of set values. As such the `m_lastValue` will always "exist", and calling visit is safe to do so.

maraisr committed:
The lastValue should be stored GC counted
https://bugs.webkit.org/show_bug.cgi?id=284334

Reviewed by Darin Adler.

The lastValue is held on, possibly between GC phases
as such we should ensure we visit the value.

* Source/WebCore/dom/InternalObserverLast.cpp:

Canonical link: https://commits.webkit.org/287715@main
